libbpf-tools/xfsslower.bpf.c

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#define BPF_NO_PRESERVE_ACCESS_INDEX
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "xfsslower.h"
+
+#define NULL    0
+
+const volatile pid_t targ_tgid = 0;
+const volatile __u64 min_lat = 0;
+
+struct piddata {
+	u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline int
+probe_entry(struct file *fp, loff_t s, loff_t e)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.start = s;
+	piddata.end = e;
+	piddata.fp = fp;
+	if (piddata.fp)
+		bpf_map_update_elem(&start, &pid, &piddata, 0);
+	return 0;
+}
+
+SEC("kprobe/xfs_file_read_iter")
+int BPF_KPROBE(kprobe__xfs_file_read_iter, struct kiocb *iocb,
+	       struct iov_iter *to)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return 0;
+}
+
+SEC("kprobe/xfs_file_write_iter")
+int BPF_KPROBE(kprobe__xfs_file_write_iter, struct kiocb *iocb,
+	       struct iov_iter *from)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return probe_entry(fp, start, 0);
+}
+
+SEC("kprobe/xfs_file_open")
+int BPF_KPROBE(kprobe__xfs_file_open, struct inode *inode, struct file *file)
+{
+	return probe_entry(file, 0, 0);
+}
+
+SEC("kprobe/xfs_file_fsync")
+int BPF_KPROBE(kprobe__xfs_file_fsync, struct file *file, loff_t start,
+	       loff_t end, int fdatasync)
+{
+	return probe_entry(file, start, end);
+}
+
+static __always_inline int probe_exit(struct pt_regs *ctx, char type)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	u64 end_ns = bpf_ktime_get_ns();
+	struct piddata *piddatap;
+	struct event event = {};
+	struct file *fp = NULL;
+	struct qstr qs = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+	u64 delta_us;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddatap = bpf_map_lookup_elem(&start, &pid);
+	if (!piddatap)
+		return 0;    /* missed entry */
+
+	delta_us = (end_ns - piddatap->ts) / 1000;
+	bpf_map_delete_elem(&start, &pid);
+
+	if ((s64)delta_us < 0 || delta_us <= min_lat * 1000)
+		return 0;
+
+	fp = piddatap->fp;
+	qs = BPF_CORE_READ(fp, f_path.dentry, d_name);
```

Comment:
what you did here w.r.t. piddatap->fp is right, `BPF_CORE_READ(piddatap, fp, f_path.dentry, d_name)` isn't supposed to work, sorry for misleading you initially. The reason is that `struct piddata *` is local type available only in your BPF application, not in kernel. There is both no need to relocate access to struct piddata, as well as it's going to result in invalid CO-RE relocation (because it can't be matched against kernel type). See error message in https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L4283

But, I think it's dangerous to read entire qstr like you do here. If the size of this struct changes, you'll read only part of it, potentially. Given you need both qstr.len and qstr.name, it's better to read them individually:

```
dentry = BPF_CORE_READ(fp, f_path.dentry);
qs_len = BPF_CORE_READ(dentry, d_name.len);
qs_name_ptr = BPF_CORE_READ(dentry, d_name.name);
bpf_probe_read_str(&event.file, sizeof(event.file), qs_name_ptr);
```

Note you don't need bpf_core_read_str() because you are using direct pointer to array off bytes, no need for relocation there.

Reply:
Get it, thanks for the details.